kombu/transport/redis_cluster.py

```diff
@@ -332,7 +335,11 @@ def _brpop_read(self, **options):
                 self.client.reinitialize_counter = 0
             else:
                 self.client.nodes_manager.update_moved_exception(e)
+            self.connection.cycle._unregister((self, self.client, conn, 'BRPOP'))
             raise Empty()
+        except Exception as e:
+            self.connection.cycle._unregister((self, self.client, conn, 'BRPOP'))
+            raise
```

Comment:
i do not believe it will handle failover or scale in/out smoothly, like I said before you have to migrate many failover and scaling in/out handling logic from redis-cluster code. And it looks really dangerous to me that you write code without any automated test that is really important for our overall logic

Reply:
I run test cases manually https://github.com/sendbird/kombu/blob/master/t/integration/test_redis_cluster.py but it's really hard to create testcase for exceptional situtations like cluster migration or connection errors.

Reply:
I'll create another PR to make it able to use RedisCluster in soda's codebase

Reply:
When I saw the code, I think using SodaRedisCluster is not enough since this transport does not use RedisCluster's execute_command but use `Redis` client and its socket.

Reply:
you should move its exception handling logic from redis-py to here

Reply:
please check this code https://github.com/sendbird/redis-py/blob/master/redis/cluster.py#L1942

Reply:
> I run test cases manually https://github.com/sendbird/kombu/blob/master/t/integration/test_redis_cluster.py but it's really hard to create testcase for exceptional situtations like cluster migration or connection errors.

I understand the restriction and hardness of creating test case for these exception, but I believe there are some ways that we can do. it looks like there are some tries to test invalid connection situation in [code base](https://github.com/sendbird/kombu/blob/master/t/integration/test_redis.py#L32) already, can you try it once?

Reply:
I think correct code is at https://github.com/sendbird/redis-py/blob/master/redis/cluster.py#L1173 , I'll copy these

Reply:
I referred pipeline's code deliberately since the code block you referred assume retry when it faces error. Because our code does not retry when exception is raised, we should approach differently

Reply:
We don't need to retry on error handler, as next `BRPOP` will be called on next poll

Reply:
Yes that's why I mentioned you should approach differently rather than copying code from `RedisCluster`

Reply:
Current exception handling codes are not direct copy of redis-py, but it has some changes for next retry.